support inplace test coverage report

HeinrichAD opened the pull request:
Add more flexibility to manipulate and positioning the coverage report.

Changes:

- option to hide page title
- possibility to chose where to add the coverage page content:
    Instead of overwriting a possible existing coverage markdown page, check if the page already exists.
    If not, just add the "new" coverage page as before.
    If the page already exists, check for the placeholder `{{mkdocs-coverage}}` and add the coverage page content there, or if the placeholder does not exists add the coverage page content at the end of the existing page.

Commit:
support inplace test coverage

Add more flexibility to manipulate and position the coverage report.

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -53,20 +48,36 @@ def on_files(self, files: Files, config: MkDocsConfig, **kwargs: Any) -> Files:
         """
-        page_name = self.config.page_name
-        page_path: str
-        if page_name is not None:
-            warnings.warn(
-                "The 'page_name' configuration option is deprecated and will be removed in a future release. "
-                "Use the 'page_path' configuration option instead.",
-                DeprecationWarning,
-                stacklevel=1,
-            )
-            page_path = page_name
```
Unless the breaking change and major version bump is required for your changes here, please revert this removal of the deprecated `page_name` option everywhere relevant 🙂

HeinrichAD replied:
Sure. It make sense to remove that in the next major update instead of now.

Instead of reverting it completely, I used `mkdocs.config.config_options.Deprecated` which results in the following difference. Please tell me if you prefer the previous behavior.

Before:

```txt
$ mkdocs serve
INFO    -  Building documentation...
INFO    -  DeprecationWarning: The 'page_name' configuration option is deprecated and will be removed in a future release. Use the 'page_path' configuration option instead.
             File "/home/****/dev/python-template/.venv/lib/python3.10/site-packages/mkdocs/plugins.py", line 566, in run_event
               result = method(item, **kwargs)
             File "/home/****/dev/python-template/.venv/lib/python3.10/site-packages/mkdocs_coverage/plugin.py", line 58, in on_files
               warnings.warn(
INFO    -  Cleaning site directory
...
```

Now:
```txt
$ mkdocs serve
WARNING -  Config value 'plugins': Plugin 'coverage' option 'page_name': The configuration option 'page_name' has been deprecated and will be removed in a future release. Use 'page_path' instead.
INFO    -  Building documentation...
INFO    -  Cleaning site directory
...
```

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -53,20 +48,36 @@ def on_files(self, files: Files, config: MkDocsConfig, **kwargs: Any) -> Files:
+        """
+        hide_title = self.config.hide_page_title
+        if hide_title is None:
+            # by default hide title if there is no user-defined coverage page content
+            hide_title = not bool(page_content)
```
Not sure to understand this. Can you illustrate/elaborate?

HeinrichAD replied:
Maybe it was a little bit to confusing. I tried to improve the implementation to make it easier to read and maintain.

The issue is that the plugin adds the following CSS code:

```css
<style>
article h1, article > a, .md-sidebar--secondary {
    display: none !important;
}
</style>
```

This CSS code will forcibly hide the toc and page title of the coverage page. This is totally fine for the generated pages but not optimal for pre-existing coverage report page which support is added with with PR.

I decided to remove newly introduced `hide_page_title` option, but still only force to hide toc and title for automatically generated coverage pages by adding this CSS code only in that specific case.

pawamoy replied:
That makes sense, thanks!

pawamoy reviewed (changes requested):
Thanks for the PR! Not sure to understand the `hide_title` option, see comments. Also, please keep the deprecated/legacy code, unless it conflicts with your changes.

HeinrichAD committed:
revert deprecated option removal and remove new hide_page_title option

- revert removal w.r.t. deprecated page_name option but use mkdocs.config.config_options.Deprecated
- remove newly introduced hide_page_title option, but still only force to hide toc and title (css) for automatically generated coverage pages
- fix: DeprecationWarning: "To replace an existing file, call `remove` before `append`."

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -26,9 +24,10 @@
     """Configuration options for the plugin."""
 
-    page_name = Optional(MkType(str, default=None))
+    page_name = MkDeprecated(moved_to="page_path", option_type=MkOptional(MkType(str, default=None)))
```
We should reuse the same message as before:


```suggestion
    page_name = MkDeprecated(
        moved_to="page_path",
        option_type=MkOptional(MkType(str, default=None)),
        message="The 'page_name' configuration option is deprecated and will be removed in a future release. "
                "Use the 'page_path' configuration option instead."
    )
```

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -53,31 +52,30 @@ def on_files(self, files: Files, config: MkDocsConfig, **kwargs: Any) -> Files:
-            page_path = self.config.page_path
-        self.page_path = page_path
-        covindex = "covindex.html" if config.use_directory_urls else f"{page_path}/covindex.html"
+        self.page_path = self.config.page_path if self.config.page_name is None else self.config.page_name
```
Isn't `self.config.page_name` moved to `self.config.page_path` by MkDocs anyway? Can't we therefore just do:


```suggestion
        self.page_path = self.config.page_path
```

HeinrichAD replied:
Thanks for the reminder. Nice catch!

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -104,18 +101,25 @@ def on_files(self, files: Files, config: MkDocsConfig, **kwargs: Any) -> Files:
+                """,
+            )
+            return style + coverage_page_content
+        if page_content.__contains__(self.config.coverage_inplace_placeholder):
```
`__contains__`? 😅 Why not `in`?

```suggestion
        if self.config.coverage_inplace_placeholder in page_content:
```

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -26,9 +24,10 @@
+    page_name = MkDeprecated(moved_to="page_path", option_type=MkOptional(MkType(str, default=None)))
     page_path = MkType(str, default="coverage")
     html_report_dir = MkType(str, default="htmlcov")
+    coverage_inplace_placeholder = MkType(str, default="{{mkdocs-coverage}}")
```
We have to use something different, as this could be interpreted by mkdocs-macros and similar plugins using Jinja syntax. WDYT of an HTML comment instead?


```suggestion
    coverage_inplace_placeholder = MkType(str, default="<!-- mkdocs-coverage -->")
```

HeinrichAD replied:
I am fine with that. I just checked it and it is working. It is also possible to use `::mkdocs-coverage::` or what ever is preferred. Since it is an option the users can always adapt it to their needs.

What do you prefer:

- `{{mkdocs-coverage}}`
- `::mkdocs-coverage::`
- `<!-- mkdocs-coverage -->`

Should I implement the last one?

pawamoy replied:
Yeah lets go with the HTML comment 😄 Thank you!

pawamoy reviewed (changes requested):
Thanks for the update!

HeinrichAD committed:
implement feedback (code improvement)

- MkDeprecated already "moved" the content of page_name to moved_to, hence we do not need to do that manually
- use `str in var` instead of `var.__contains__(str)`

HeinrichAD committed:
change in-place coverage default placeholder

Switch from curly brackets to html comment to avoid conflicts with other plugins like mkdocs-macros etc.

pawamoy commented on src/mkdocs_coverage/plugin.py:
```diff
@@ -26,19 +24,20 @@
+    )
     page_path = MkType(str, default="coverage")
     html_report_dir = MkType(str, default="htmlcov")
+    coverage_inplace_placeholder = MkType(str, default="<!-- mkdocs-coverage -->")
```
OK lets just rename this to a shorter `placeholder`, document it, and we'll be good to go, WDYT 🙂?

pawamoy replied:
Let me take care of it :+1:

pawamoy committed:
fixup

pawamoy committed:
fixup

pawamoy committed:
fixup

pawamoy committed:
fixup

pawamoy commented:
Sorry for the noise added to the PR, lets just have CI pass.

pawamoy commented:
Well, everything passes locally, it just needs a `copier update`, but I'll do that in main once it's merged. Thanks a lot for the PR!
